### DISCUSSION TITLE
Self Hosting Edge Functions

### DISCUSSION THREAD

**Discussion by Kade-Powell** [supabase/supabase#6851] [General]
Hoping to get answered by someone from SB Dev team. First, for some context, my company wants everything hosted internal on our infrastructure. I’m trying to provide a way for people to turn up new apps more easily and with less code - which is what led me here. So the other parts of supabase are great (minus the lack of official kubernetes support) the only missing piece for us is the edge functions. So when invoking a function on  supabase hosting you simply call functions.supabase URL via post. Or the sdk. However in studio when you self host, the functions are not even an icon on the side bar. From reading blog posts I see self hosting functions were planned at some point but I was hoping to see how soon or if the back-end functionality is there etc. I know platforms like appwrite can do them but there platform is based on noSQL databases and that won’t work for us either. So I’m trying to see how I can provide a similar experience of the cloud hosted version to developers here - if successful I will be able to secure a support contract with supabase, if willing. 

Thanks for any help.

**Comment by jaitaiwan** [comment]
Hey @Kade-Powell,

Edge functions in supabase are done using the Deno Deploy platform. In front of every edge function that's deployed is a functions-relay: https://github.com/supabase/functions-relay

Hopefully this helps?

**Reply by Kade-Powell** [reply]
Are there any blog post describing how to achieve something similar with a platform like openFaaS?

**Reply by SteveChurch** [reply]
@Kade-Powell did you ever find any useful resources / documentation around this? I'm going to have a play with the `functions-replay` that @jaitaiwan has mentioned above. 

EDIT: Right, now I see, Supabase hosts Edge Functions on Deno Deploy

**Reply by Kade-Powell** [reply]
No I didn't. Although you could do it with a little custom setup using something like openFaaS. I haven't seen anyone do it though

**Reply by SteveChurch** [reply]
Thanks, I'll take a look. I am hoping to keep everything in one place, ideally I want my edge functions to run in my rocker environment.

**Comment by SteveChurch** [comment]
@jaitaiwan so were saying for every function we want to deploy to Deno Deploy it needs a a Functions Gateway container running to point to it? 

What I can't seem to understand is, does every function run in a separate project in Deno Deploy or does Supabase bundle additional code to handle URIs for each endpoint? 

Edit: 

'''

When a request comes in to an Edge Function you deploy on Supabase, it first lands at a “Relay”. The Relay acts as an API gateway, authenticating the JWT, and providing some additional functionality like logging and rate-limiting. Upon receiving the incoming request, the Relay retrieves the metadata for your Function, and uses it to construct a unique identifier (”Deployment ID”) for the Function. This Deployment ID then gets passed over to the Deno Deploy platform, which takes care of invoking your Function securely, and passing your output back to the Relay, which in turn passes it back to your end-user.
''' 

Hmm.
